flink-runtime/src/test/java/org/apache/flink/runtime/resourcemanager/slotmanager/DeclarativeSlotManagerTest.java

```diff
@@ -0,0 +1,1092 @@
+
+			slotManager.registerTaskManager(taskManagerConnection, slotReport);
+
+			assertThat("The number registered slots does not equal the expected number.", slotManager.getNumberRegisteredSlots(), is(2));
+
+			assertNotNull(slotTracker.getSlot(slotId1));
+			assertNotNull(slotTracker.getSlot(slotId2));
+		}
+	}
+
+	/**
+	 * Tests that un-registration of task managers will free and remove all registered slots.
+	 */
+	@Test
+	public void testTaskManagerUnregistration() throws Exception {
+		final TaskExecutorGateway taskExecutorGateway = new TestingTaskExecutorGatewayBuilder()
+			.setRequestSlotFunction(tuple6 -> new CompletableFuture<>())
+			.createTestingTaskExecutorGateway();
+		final TaskExecutorConnection taskManagerConnection = createTaskExecutorConnection(taskExecutorGateway);
+		final ResourceID resourceId = taskManagerConnection.getResourceID();
+
+		final SlotID slotId1 = new SlotID(resourceId, 0);
+		final SlotID slotId2 = new SlotID(resourceId, 1);
+		final SlotReport slotReport = new SlotReport(Arrays.asList(createAllocatedSlotStatus(slotId1), createFreeSlotStatus(slotId2)));
+
+		final ResourceRequirements resourceRequirements = createResourceRequirementsForSingleSlot();
+
+		final DefaultSlotTracker slotTracker = new DefaultSlotTracker();
+
+		try (DeclarativeSlotManager slotManager = createDeclarativeSlotManagerBuilder()
+			.setSlotTracker(slotTracker)
+			.buildAndStartWithDirectExec()) {
+
+			slotManager.registerTaskManager(taskManagerConnection, slotReport);
+
+			assertEquals("The number registered slots does not equal the expected number.", 2, slotManager.getNumberRegisteredSlots());
+
+			DeclarativeTaskManagerSlot slot1 = slotTracker.getSlot(slotId1);
+			DeclarativeTaskManagerSlot slot2 = slotTracker.getSlot(slotId2);
+
+			assertSame(SlotState.ALLOCATED, slot1.getState());
+			assertSame(SlotState.FREE, slot2.getState());
+
+			slotManager.processResourceRequirements(resourceRequirements);
+
+			assertSame(SlotState.PENDING, slot2.getState());
+
+			slotManager.unregisterTaskManager(taskManagerConnection.getInstanceID(), TEST_EXCEPTION);
+
+			assertEquals(0, slotManager.getNumberRegisteredSlots());
+		}
+	}
+
+	/**
+	 * Tests that a slot request with no free slots will trigger the resource allocation.
+	 */
+	@Test
+	public void testSlotRequestWithoutFreeSlots() throws Exception {
+		final ResourceManagerId resourceManagerId = ResourceManagerId.generate();
+
+		final ResourceRequirements resourceRequirements = createResourceRequirementsForSingleSlot();
+
+		CompletableFuture<WorkerResourceSpec> allocateResourceFuture = new CompletableFuture<>();
+		ResourceActions resourceManagerActions = new TestingResourceActionsBuilder()
+			.setAllocateResourceConsumer(allocateResourceFuture::complete)
+			.build();
+
+		try (SlotManager slotManager = createSlotManager(resourceManagerId, resourceManagerActions)) {
+
+			slotManager.processResourceRequirements(resourceRequirements);
+
+			allocateResourceFuture.get();
```

Comment:
Maybe check that the right sized resource is requested.

Reply:
The size of the requested resource is not really dependent on the resource requirements; the only thing that is checked (by the TaskExecutorManager) is that the slots of said resource can fulfill the requirements.
This is admittedly not tested, but belongs into the TaskExecutorManagerTest imo.

Reply:
Ok sounds good.